075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,490 @@
+
+The FSM is initialized based on the `state` field.
+
+The `Kafka.status.autoRebalance.modes[add-brokers|remove-brokers].brokers` list is filled or updated based on its current content and what's coming from `removeNodes` and `addedNodes`.
+
+Check the `removeNodes`:
+
+  * if empty
+    * no further action and stay with the current `Kafka.status.autoRebalance.modes[remove-brokers].brokers` list (which could mean being empty/not existing too).
+  * if not empty
+    * update the `Kafka.status.autoRebalance.modes[remove-brokers].brokers` by producing a consistent list with its current content and what is in the `removeNodes` list. This cover the case that while rebalancing scale down running, the user requests an update on an already running one.
+
+Check the `addedNodes`:
+
+  * if empty
+    * no further action and stay with the current `Kafka.status.autoRebalance.modes[add-brokers].brokers` list (which could mean being empty/not existing too)
+  * if not empty
+    * update the `Kafka.status.autoRebalance.modes[add-brokers].brokers` by producing a consistent list with its current content and what is in the `addedNodes` list. This cover the case that while rebalancing scale down running, the user requests a new scale up or an update on an already queued one. Or even the case that while rebalancing scale up running, the user requests an update on an already running one.
+
+During the FSM processing, the `Kafka.status.autoRebalance.modes[add-brokers|remove-brokers].brokers` list is compared with the corresponding `KafkaRebalance.spec.brokers` list for the current rebalancing to determine if there were changes.
+Those changes could mean several use cases: a new scale down/up requested while another one is running, a scale up requested while a scale down is running and so on.
+
+Let's see what happens during the auto-rebalancing process when the FSM starts from a specific state and depending on the brokers list updates.
+Remember that starting a rebalancing always means creating a corresponding "actual" `KafkaRebalance` customer resource from the "template" one and applying the `strimzi.io/rebalance-auto-approval: true` annotation.
+
+#### Idle
+
+This is when `Kafka.status.autoRebalance` is empty or `Kafka.status.autoRebalance.state` is `Idle` because of a previous successfully completed or failed auto-rebalancing.
+If failed, the `Kafka.status.autoRebalance.state.error` contains an error message.
+
+From this state:
+
+* if the previous auto-rebalancing was completed successfully, the user can trigger a new auto-rebalancing by scaling down/up the cluster again.
+* if the previous auto-rebalancing failed, the user has to deal with the rebalancing on their own (with the usual `KafkaRebalance` custom resource manual creation). If they are not interested to fix it manually, on the next scaling down/up the cluster, the auto-rebalancing process will be triggered again, cleaning the error.
```

Comment:
This seems a bit fishy. It sounds like the state would:
* either disappear right in the next reconciliation (in which case should we just rely on logging) 
* or it will be there until the next auto-rebalance
* or in the worse case it will trigger the auto-rebalance again and again for already completed nodes?

Reply:
> either disappear right in the next reconciliation (in which case should we just rely on logging)

Yeah. When I say "on the next scaling down/up the cluster, the auto-rebalancing process will be triggered again, cleaning the error", it could easily happen automatically on the next reconciliation due to a scale down (because the scale up won't trigger again, pods are running already). And in case of more errors this could trigger the loop (as your third point).

I guess that at this point the only easy solution is leaving the user to deal with the error? The autorebalance status could stay in error there (even forever) just to show that last rebalancing failed but on next reconciliation the FSM (because of the error) would not handle a new autorebalancing (this would stop the infinite triggering). The question what's the best UX for the user to clear this error, to be sure that in a future scale up/down, the autorebalancing is processed? If they go the manual way, the rebalance is successful and the related scale as well, how to unlock the operator for a future new autorebalancing?

Reply:
My take is that there is no perfect option. I would probably treat the status here as internal only and have the users to rely on logs only? That way, you can simple clear it on error and be done with it. If you don't clear it, it will IMHO be too chaotic:
* You will need to know what to ignore when next auto-rebalance is needed
* But that means It is not a reliable source of information for the user because next auto-rebalance event would wipe the status anyway
* The scaling might be a rare event ... I'm not completely comfortable with some scale-up rebalance error being in the status several years after it happened because that was the last scaling.
* For scale-down, the non-empty nodes will still be there so a new scale-won will be triggered again after few seconds. So the error again does not have much value for the user and logs will be the way to go..

Should we consider storing this in some ConfigMap instead of in `.status` to make it more clear it is internal information only? I'm fine with it in the status, but this would _hide_ it from the user if we would think it would be too confusing to have in the status but mostly as internal information only.

Reply:
I would prefer having the info in the `Kafka.status` because this is the usual UX we provide to users to check what's up with the Kafka cluster. Also, it would provide information during the rebalancing about what's happening and which brokers are involved. On the other hand, the user could also take a look on the corresponding `KafkaRebalance` resource to know how things are going.
Now, if we wipe out the status when the rebalancing ends with error, how the user is supposed to know that there was a problem? For scale down, user should see the scaling being blocked but for scale up there is no way to know that the subsequent rebalance failed. At the same time, taking the Idle state and the error there could drive the issues you mention. Of course, as said before a way would be looking at the `KafkaRebalance` resource.

Another point related to this would be ... who should delete the `KafkaRebalance` resource at the end of the rebalancing (completed or failed)? The user or the operator? Leaving it at the user would allow him to have a place where errors could be checked? If done the operator, there is really no way for the user to know how rebalancing went.

Reply:
Yeah, but ... as I said ...
* You need to remove the error at some point. You do not want to keep it forever as scaling events might be rare
* For scale-down, you would trigger another right away and anyway override the error message

So I'm not sure this works for informing users and IMHO you still need to rely on the regular log.

Reply:
> For scale-down, you would trigger another right away and anyway override the error message

Thinking more maybe this is not true. The FSM could avoid starting a new rebalancing if the state is `Idle` and the `Kafka.status.autoRebalance.state.error` field is filled. It's waiting for the user to "clean/reset". The problem is what's the best UX for doing this? The user should delete the `KafkaRebalance` for example? Or what if the user thinks "ok don't do any autorebalance anymore, I will take care of it manually".

Reply:
I think that just over complicates it. You need to figure out how the user would _reset it_ and make sure the user knows that they have to reset it. And until then it will be blocked just collecting scaling up and down events. I don't think that makes sense as a workflow but it likely also adds a lot complexity? To me it makes more sense to rely on logs and possibly some metrics

Reply:
We are on the same page, I see all these difficulties but what's not clear to me in your comment is how you envisage the way for the operator to figure out that it has to stop auto rebalancing on next reconciliation, mostly in case of scale down where nodes to remove are still there, the Kafka.spec.cruiseControl.autoRebalance is still there and it injects the mechanism to start if the status is wiped out/empty at the beginning of the next reconciliation.

Reply:
For scale down, it will essentially never stop ....
* User requests the scale down
* Operator blocks it
* Auto-rebalance to remove the nodes is triggered
* ...
* Auto-rebalance fails, you log the error, delete the `KafkaRebalance` resource
* Next reconciliation starts
* The nodes are still requested for scale-down and still blocked
* You trigger another auto-rebalance to remove them
* ...
* This way it proceeds either until it succeeds (it the error was something temporary) or until the user handles the issue 🤷 

For scale-up, this would be different I guess ...
* User requests scale-uo
* Operator adds the nodes
* Auto-rebalance to add the nodes is triggered
* ...
* Auto-rebalance fails, you log the error, delete the `KafkaRebalance` resource ... and forget it

That would be how I would consider handling it.

----

I think recovering from the scale-up rebalance failure would be much harder ... are the nodes still there? Is the scale-up still needed? I think you would easily get lost in the various combinations and would need to rely on the user to take an action that is not obvious. So I'm not sure it is worth it.

For scale-down, it is much easier because you always care only about the nodes for which the scale-down was reverted. So you can easily recover if the error is temporary. And if the error is permanent, the nodes will be still there which makes it much more likely that the user will jump in and fix it.

Reply:
Of course, this approach is the easier one I would say but I can't see any which is more user friendly. The only thing which is bothering me is that we leave the user no idea what's going on in terms of autorebalancing in the Kafka CR status but just log messages (even because we delete the `KafkaRebalance` which remains the last source of truth for any error). For now I am going to update the proposal this way and see what the others think, if someone comes with a better approach. Thanks!

Reply:
Yeah, I do not think it is more user friendly. But I think it is not significantly less user friendly and is easier to code.